### PR TITLE
january cleanup

### PR DESCRIPTION
**Pull request by exaexa**

### CODE REVIEW

**Comment by exaexa** [comment]
The remaining issue for 1.0 is the #22 which will need a bit of support from DataStructures. I might make a temporary (slower) version too, will see. I think this is ready now tho.

File: src/linear_value.jl [hunk 1]
```diff
@@ -59,42 +61,63 @@ Base.:*(a::Real, b::LinearValue) = b * a
...
+"sparse" representations of linear combinations, it computes a "merged" linear
+combination of 2 values added together.
+
+Zeroes are not filtered out.
+"""
+function add_sparse_linear_combination(
+    a_idxs::Vector{Int},
+    a_weights::Vector{T},
+    b_idxs::Vector{Int},
+    b_weights::Vector{T},
```
Comment (stelmo) [inline src/linear_value.jl:77]: yes!

File: src/quadratic_value.jl [hunk 2]
```diff
@@ -90,58 +90,96 @@ Internal helper for co-lex ordering of indexes.
...
+Helper function for multiplying two [`LinearValue`](@ref)-like objects to make
+a [`QuadraticValue`](@ref)-like object. This computes and merges the product.
+
+Zeroes are not filtered out.
+"""
+function multiply_sparse_linear_combination(
+    a_idxs::Vector{Int},
+    a_weights::Vector{T},
+    b_idxs::Vector{Int},
+    b_weights::Vector{T},
```
Comment (stelmo) [inline src/quadratic_value.jl:162]: yes^2

File: src/tree.jl [hunk 3]
```diff
@@ -113,6 +113,6 @@ elements of type specified by the 3rd argument. (This needs to be specified
 explicitly, because the typesystem generally cannot guess the universal type
 correctly.)
 """
-tree_map(x::Tree, f, ::Type{T}) where {T} = Tree{T}(k => tree_map(v, f, T) for (k, v) in x)
+tree_map(f, x::Tree, ::Type{T}) where {T} = Tree{T}(k => tree_map(f, v, T) for (k, v) in x)
```
Comment (stelmo) [inline src/tree.jl:116]: good, more in line with map functions I think

**Review by stelmo — APPROVED** [review]
great!

**Comment by exaexa** [comment]
ok well actually let's add at least a primitive version of the tree_zip... :D

File: src/tree.jl (hunk 3 above)
Comment (exaexa) [reply src/tree.jl:116]: yeah also allows the `do` notation

```
map(mytree, Int) do c
  whatever(c.value .....)
 end
```

File: docs/src/1-metabolic-modeling.jl [hunk 4]
```diff
@@ -374,8 +378,52 @@ c[:exchanges][:production_is_zero] = C.Constraint(c.exchanges.biomass.value, 0)
...
+# function to the contents of several trees, "zipping" them over the same keys
+# in the structure. This is vaguely similar but otherwise not related to the
+# `zip` from Julia base (similarly, ConstraintTrees.jl have their own specific
+# `map`).
+#
+# In practice, this allows you to create combined trees with various nice
+# properties very quickly. For example, you can find how much the values have
+# changed between our two communities:
+
+C.zip((x, y) -> y - x, result, result_with_more_oxygen, Float64)
```
Comment (stelmo) [inline docs/src/1-metabolic-modeling.jl:399]: this is so cool

File: src/constraint_tree.jl [hunk 5]
```diff
@@ -211,11 +212,26 @@ function variables(; keys::AbstractVector{Symbol}, bounds = nothing)
...
+`makebound` function, which converts a given tree element's value into a bound
+for the corresponding variable.
+"""
+function variables_for(makebound, ts::Tree)
+    var_idx = 0
+    map(ts, Constraint) do x
+        var_idx += 1
+        variable(idx = var_idx, bound = makebound(x))
+    end
+end
```
Comment (stelmo) [inline src/constraint_tree.jl:233]: where is this used?

File: src/tree.jl [hunk 6]
```diff
@@ -113,6 +113,26 @@ elements of type specified by the 3rd argument. (This needs to be specified
...
+)
+
+zip(f, x::Tree, y::Tree, z::Tree, ::Type{T}) where {T} = Tree{T}(
+    k => zip(f, x[k], y[k], z[k], T) for
+    k in intersect(SortedSet(keys(x)), SortedSet(keys(y)), SortedSet(keys(z)))
+)
+
+zip(f, x, y, ::Type) = f(x, y)
+
+zip(f, x, y, z, ::Type) = f(x, y, z)
```
Comment (stelmo) [inline src/tree.jl:138]: cool

**Review by stelmo — APPROVED** [review]

File: src/constraint_tree.jl (hunk 5 above)
Comment (stelmo) [reply src/constraint_tree.jl:233]: i.e. when would I want to use it?

Comment (exaexa) [reply src/constraint_tree.jl:233]: https://github.com/LCSB-BioCore/COBREXA.jl/pull/814/files#diff-4073272f80aada4226d99a069713c2e52b6abeb93d66e1124c7f87e3cc9f4abbR48-R54

File: src/tree.jl (hunk 3 above)
Comment (exaexa) [reply src/tree.jl]: ok this is demonstrated in docs now, all ok.

**Pull request merged**